File: library/src/main/java/tw/shounenwind/plurkconnection/callbacks/ApiJsonElementCallback.java
package tw.shounenwind.plurkconnection.callbacks;

import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

import okhttp3.Response;
import okhttp3.ResponseBody;
import tw.shounenwind.plurkconnection.responses.ApiResponseJsonElement;

public abstract class ApiJsonElementCallback extends BasePlurkCallback<ApiResponseJsonElement> {

    @Override
    public void runResult(Response response) throws Exception {
        super.runResult(response);
        ResponseBody body = null;
        ApiResponseJsonElement result;
        try {
            body = response.body();
            JsonElement jsonElement = new JsonParser().parse(body.charStream());
            result = new ApiResponseJsonElement(response.code(), jsonElement);
        } finally {
            if (body != null) {
                body.close();
            }
        }
        onSuccess(result);
    }
}

File: library/src/main/java/tw/shounenwind/plurkconnection/callbacks/ApiJsonElementCallback.kt
package tw.shounenwind.plurkconnection.callbacks

import com.google.gson.JsonParser
import okhttp3.Response
import okhttp3.ResponseBody
import tw.shounenwind.plurkconnection.responses.ApiResponseJsonElement

abstract class ApiJsonElementCallback : BasePlurkCallback<ApiResponseJsonElement>() {

    @Throws(Exception::class)
    override fun runResult(response: Response) {
        super.runResult(response)
        var body: ResponseBody? = null
        val result: ApiResponseJsonElement
        try {
            body = response.body()
            val jsonElement = JsonParser().parse(body!!.charStream())
            result = ApiResponseJsonElement(response.code(), jsonElement)
        } catch (e: Exception) {
            throw e
        } finally {
            body?.close()
        }
        onSuccess(result)
    }
}
